core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -0,0 +1,416 @@
+         throw log.getConflictsStateTransferInProgress(cacheName);
+
+      if (!streamInProgress.compareAndSet(false, true))
+         throw log.getConflictsAlreadyInProgress();
+
+      try {
+         return StreamSupport
+               .stream(new ReplicaSpliterator(topology), false)
+               .filter(filterConsistentEntries());
+      } finally {
+         streamInProgress.compareAndSet(true, false);
+      }
+   }
+
+   @Override
+   public boolean isConflictResolutionInProgress() {
+      return streamInProgress.get();
+   }
+
+   @Override
+   public void resolveConflicts() {
+      resolveConflicts(installedTopology);
+   }
+
+   @Override
+   public void resolveConflicts(CacheTopology topology) {
+      LocalizedCacheTopology localizedTopology;
+      if (topology instanceof LocalizedCacheTopology) {
+         localizedTopology = (LocalizedCacheTopology) topology;
+      } else {
+         localizedTopology = distributionManager.createLocalizedCacheTopology(topology);
+      }
+      doResolveConflicts(localizedTopology);
+   }
+
+   private void doResolveConflicts(final LocalizedCacheTopology topology) {
+      final Set<Address> preferredPartition = new HashSet<>(topology.getCurrentCH().getMembers());
+
+      if (trace) log.tracef("Attempting to resolve conflicts.  All Members %s, Installed topology %s, Preferred Partition %s",
+            topology.getMembers(), topology, preferredPartition);
+
+      final Phaser phaser = new Phaser(1);
+      getConflicts(topology).forEach(conflictMap -> processConflict(phaser, conflictMap, topology, preferredPartition));
+      phaser.arriveAndAwaitAdvance();
+   }
+
+   private void processConflict(Phaser phaser, Map<Address, CacheEntry<K, V>> conflictMap,
+                                                LocalizedCacheTopology topology, Set<Address> preferredPartition) {
+      stateTransferExecutor.execute(() -> {
+         phaser.register();
+         if (trace) log.tracef("Conflict detected %s", conflictMap);
+
+         Collection<CacheEntry<K, V>> entries = conflictMap.values();
+         final K key = entries.iterator().next().getKey();
+         Address primaryReplica = topology.getDistribution(key).primary();
+
+         List<Address> preferredEntries = conflictMap.entrySet().stream()
+               .map(Map.Entry::getKey)
+               .filter(preferredPartition::contains)
+               .collect(Collectors.toList());
+
+         // If only one entry exists in the preferred partition, then use that entry
+         CacheEntry<K, V> preferredEntry;
+         if (preferredEntries.size() == 1) {
+            preferredEntry = conflictMap.remove(preferredEntries.get(0));
+         } else {
+            // If multiple conflicts exist in the preferred partition, then use primary replica from the preferred partition
+            // If not a merge, then also use primary as preferred entry
+            // Preferred is null if no entry exists in preferred partition
```

Comment:
This is a bit inconsistent; if there is a single conflict in the preferred partition (e.g. because the entry on primaryReplica has been removed completely), you use that; if you have two other nodes (different from primary replica) with conflict, you take the value from primary replica which was not marked as a conflict before. Not sure if you can solve this without tombstones, though :-/ But at least document that.